eventsub-websocket/src/main/java/com/github/twitch4j/eventsub/socket/TwitchEventSocket.java

```diff
@@ -0,0 +1,662 @@
+                this.connect(); // will use the latest url
+                break;
+
+            case NOTIFICATION:
+                final EventSubEvent event;
+                try {
+                    event = Objects.requireNonNull(payload.getParsedEvent());
+                } catch (Exception e) {
+                    log.error("EventSub-WS failed to parse notification event data: " + payload.getEventData(), e);
+                    return;
+                }
+
+                try {
+                    eventManager.publish(event);
+                } catch (Exception e) {
+                    log.error("Non-T4J event consumer threw exception while handling EventSocket notification: " + event, e);
+                }
+                break;
+
+            case REVOCATION:
+                final EventSubSubscription revoked = payload.getSubscription();
+                if (unsubscribeNoHelix(revoked) != null) {
+                    log.debug("Removed revoked EventSub-WS subscription {}", revoked);
+
+                    try {
+                        eventManager.publish(payload.getSubscription());
+                    } catch (Exception e) {
+                        log.error("Non-T4J event consumer threw exception while handling revoked EventSocket subscription", e);
+                    }
+                } else {
+                    log.warn("Failed to identify revoked EventSub-WS subscription {}", revoked);
+                }
+                break;
+        }
+    }
+
+    private boolean createSub(OAuth2Credential credential, EventSubSubscription oldSub, EventSubSubscription newSub) {
+        OAuth2Credential token = credential != null ? credential : defaultToken;
+        try {
+            SubscriptionWrapper sub = SubscriptionWrapper.wrap(
+                api.createEventSubSubscription(getAuthToken(token), newSub)
+                    .execute()
+                    .getSubscriptions()
+                    .get(0)
+            );
+            subscriptions.put(sub, sub);
+            log.debug("EventSub-WS successfully created subscription {}", sub);
+            if (token != null) tokenByTopic.put(sub, token);
+            eventManager.publish(new EventSocketSubscriptionSuccessEvent(sub, this));
+            return true;
+        } catch (Exception e) {
+            log.error("Failed to create EventSub-WS subscription {}", newSub, e);
+
+            // skip retry on confirmed bad subscriptions
+            boolean retry = true;
+            if (avoidRetryFailedSubscription && e instanceof HystrixRuntimeException && e.getCause() instanceof ContextedRuntimeException) {
+                ContextedRuntimeException cause = (ContextedRuntimeException) e.getCause();
+                String status = String.valueOf(cause.getFirstContextValue("errorStatus"));
+                try {
+                    int code = Integer.parseInt(status);
+                    if (code >= 400 && code < 500 && code != 429) {
```

Comment:
Should we retry 401's ?

Reply:
I'd say no, we have the event to inform the user about the failure and they either get a valid token to resubscribe or ignore it (no retries in this case).